Bump Swift version to 4.0

tkremenek opened the pull request:
rdar://problem/31899950

tkremenek committed:
Bump Swift version to 4.0

tkremenek commented:
@swift-ci clean test

tkremenek committed:
Adjust to return compatibility version in Swift 3 mode.

tkremenek commented:
@swift-ci clean test

erg commented:
@tkremenek The other place to change is in ``build-script``:

```python
    parser.add_argument(
        "--swift-user-visible-version",
        help="User-visible version of the embedded Swift compiler",
        type=arguments.type.swift_compiler_version,
        default="3.1", # change to 4.0
        metavar="MAJOR.MINOR")
```

Note: The ``build-script-impl`` version flag was made into dead code by a refactoring to python by someone else months ago, but I only removed it last week upon discovering it was dead.

tkremenek committed:
Bump default version in build script.

tkremenek commented:
@swift-ci clean test

jrose-apple reviewed (changes requested):
For now, we need to continue building the standard library and overlays with Swift 3 as well.

dabrahams commented:
on Wed Apr 19 2017, Jordan Rose <notifications-AT-github.com> wrote:

> For now, we need to continue building the standard library and
> overlays with Swift 3 as well.

That's a little alarming to hear; we've been under the impression that
we could switch to building the standard library in Swift 4 mode
whenever we want.  Please explain why not, and when we can expect to be
able to flip the switch.

Thanks,
-Dave

tkremenek commented:
@dabrahams makes a good point.  The intent is we want the Standard Library and overlays building in Swift 4 mode.

tkremenek commented:
@DougGregor reminded me offline that the Standard Library/overlays can vend Swift 4 changes with `@available` markup.  Let's review outside this PR whether we need to build these libraries in Swift 4 mode.

tkremenek committed:
Force -swift-version 3 for Standard Library and overlays.

tkremenek commented:
@swift-ci clean test

tkremenek committed:
Compile all standard library as Swift 3 mode.

tkremenek commented:
@swift-ci test linux

tkremenek commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/956

@swift-ci test linux

tkremenek commented:
Please test with following pull request:
apple/swift-corelibs-xctest#188
apple/swift-corelibs-foundation#956

@swift-ci test linux

tkremenek commented:
Build failure is now in SwiftPM.  I've reached out to @ddunbar to see how to force SwiftPM to build with `-swift-version 3` for now.

tkremenek commented:
@swift-ci smoke test macOS

tkremenek commented:
@aciidb0mb3r provided a PR for SwiftPM:

https://github.com/apple/swift-package-manager/pull/1109

tkremenek commented:
Please test with following pull request:
apple/swift-corelibs-xctest#188
apple/swift-corelibs-foundation#956
apple/swift-package-manager#1109

@swift-ci test linux

tkremenek commented:
Failing tests on macOS that probably need `-swift-version 3`:

~~~
Failing Tests (15):
    Swift(macosx-x86_64) :: attr/attr_dynamic_infer.swift
    Swift(macosx-x86_64) :: attr/attr_objc_clang.swift
    Swift(macosx-x86_64) :: Serialization/Recovery/crash-recovery.swift
    Swift(macosx-x86_64) :: IDE/crashers_2/0010-reference-to-self-in-extension-init.swift
    Swift(macosx-x86_64) :: IDE/crashers_2/0011-complete-protocol-with-initializer.swift
    Swift(macosx-x86_64) :: Serialization/rdar29694978.swift
    Swift(macosx-x86_64) :: Driver/swift-version-default.swift
    Swift(macosx-x86_64) :: IDE/complete_crashes.swift
    Swift(macosx-x86_64) :: IDE/complete_dynamic_lookup.swift
    Swift(macosx-x86_64) :: IDE/complete_func_reference.swift
    Swift(macosx-x86_64) :: IDE/complete_stdlib_optional.swift
    Swift(macosx-x86_64) :: IDE/local_types.swift
    Swift(macosx-x86_64) :: IDE/merge_local_types.swift
    Swift(macosx-x86_64) :: IDE/print_usrs.swift
    Swift(macosx-x86_64) :: IDE/range_info_basics.swift
~~~

tkremenek commented:
Failing tests on Linux:

~~~
01:24:23 Failing Tests (55):
01:24:23     Swift(linux-x86_64) :: compiler_crashers_2_fixed/0041-walkToDeclPost-crash.swift
01:24:23     Swift(linux-x86_64) :: stdlib/Renames.swift
01:24:23     Swift(linux-x86_64) :: SILGen/properties.swift
01:24:23     Swift(linux-x86_64) :: Sema/complex_expressions.swift
01:24:23     Swift(linux-x86_64) :: Serialization/class-roundtrip-module.swift
01:24:23     Swift(linux-x86_64) :: Serialization/class.swift
01:24:23     Swift(linux-x86_64) :: SILGen/guaranteed_self.swift
01:24:23     Swift(linux-x86_64) :: SILGen/closures.swift
01:24:23     Swift(linux-x86_64) :: IDE/complete_stdlib_optional.swift
01:24:23     Swift(linux-x86_64) :: Constraints/diagnostics.swift
01:24:23     Swift(linux-x86_64) :: Constraints/closures.swift
01:24:23     Swift(linux-x86_64) :: compiler_crashers_fixed/26813-generic-enum-tuple-optional-payload.swift
01:24:23     Swift(linux-x86_64) :: compiler_crashers_2_fixed/0034-rdar21982453.swift
01:24:23     Swift(linux-x86_64) :: Sema/type_checker_crashers_fixed/rdar27261929.swift
01:24:23     Swift(linux-x86_64) :: IDE/crashers_2/0010-reference-to-self-in-extension-init.swift
01:24:23     Swift(linux-x86_64) :: IDE/crashers_2/0011-complete-protocol-with-initializer.swift
01:24:23     Swift(linux-x86_64) :: expr/postfix/dot/init_ref_delegation.swift
01:24:23     Swift(linux-x86_64) :: decl/subscript/subscripting.swift
01:24:23     Swift(linux-x86_64) :: decl/protocol/req/optional.swift
01:24:23     Swift(linux-x86_64) :: expr/closure/single_expr.swift
01:24:23     Swift(linux-x86_64) :: expr/closure/closures.swift
01:24:23     Swift(linux-x86_64) :: expr/closure/anonymous.swift
01:24:23     Swift(linux-x86_64) :: decl/func/default-values.swift
01:24:23     Swift(linux-x86_64) :: decl/class/override.swift
01:24:23     Swift(linux-x86_64) :: attr/attr_dynamic_infer.swift
01:24:23     Swift(linux-x86_64) :: attr/attr_dynamic.swift
01:24:23     Swift(linux-x86_64) :: Serialization/struct.swift
01:24:23     Swift(linux-x86_64) :: Serialization/objc.swift
01:24:23     Swift(linux-x86_64) :: SILGen/tuples.swift
01:24:23     Swift(linux-x86_64) :: SILGen/property_abstraction.swift
01:24:23     Swift(linux-x86_64) :: SILGen/partial_apply_generic.swift
01:24:23     Swift(linux-x86_64) :: SILGen/nested_types_referencing_nested_functions.swift
01:24:23     Swift(linux-x86_64) :: SILGen/metatype_abstraction.swift
01:24:23     Swift(linux-x86_64) :: SILGen/existential_metatypes.swift
01:24:23     Swift(linux-x86_64) :: SILGen/function_conversion.swift
01:24:23     Swift(linux-x86_64) :: SILGen/generic_closures.swift
01:24:23     Swift(linux-x86_64) :: SILGen/dynamic_self.swift
01:24:23     Swift(linux-x86_64) :: SILGen/apply_abstraction_nested.swift
01:24:23     Swift(linux-x86_64) :: Misc/expression_too_complex.swift
01:24:23     Swift(linux-x86_64) :: IRGen/existential_nested_type.swift
01:24:23     Swift(linux-x86_64) :: IDE/range_info_basics.swift
01:24:23     Swift(linux-x86_64) :: IDE/print_usrs.swift
01:24:23     Swift(linux-x86_64) :: IDE/local_types.swift
01:24:23     Swift(linux-x86_64) :: IDE/merge_local_types.swift
01:24:23     Swift(linux-x86_64) :: IDE/complete_func_reference.swift
01:24:23     Swift(linux-x86_64) :: IDE/complete_crashes.swift
01:24:23     Swift(linux-x86_64) :: Generics/deduction.swift
01:24:23     Swift(linux-x86_64) :: Driver/swift-version-default.swift
01:24:23     Swift(linux-x86_64) :: DebugInfo/WeakCapture.swift
01:24:23     Swift(linux-x86_64) :: DebugInfo/closure-arg-linetable.swift
01:24:23     Swift(linux-x86_64) :: DebugInfo/nostorage.swift
01:24:23     Swift(linux-x86_64) :: Constraints/diag_ambiguities.swift
01:24:23     Swift(linux-x86_64) :: Constraints/tuple.swift
01:24:23     Swift(linux-x86_64) :: Compatibility/attr_escaping.swift
01:24:23     Swift(linux-x86_64) :: ClangImporter/ctypes_parse.swift
01:24:23 
~~~

jrose-apple commented:
Hm. lit is supposed to be passing -swift-version explicitly to any test using most of the standard substitutions. A few of these might be deliberately bypassing that, and Driver/swift-version-default.swift is deliberately trying to test the very thing being changed.

jrose-apple commented on cmake/modules/SwiftSource.cmake:
```diff
@@ -269,7 +269,7 @@ function(_compile_swift_files
 
   # Force swift 3 compatibility mode for Standard Library and overlay.
-  if (SWIFTFILE_IS_STDLIB_CORE OR SWIFTFILE_IS_SDK_OVERLAY)
+  if (SWIFTFILE_IS_STDLIB OR SWIFTFILE_IS_SDK_OVERLAY)
```
Which files don't fit into either of these categories? Things like StdlibUnittest?

tkremenek commented:
Some of the tests that are failing use  `%target-swift-ide-test`.  I do not believe that lit passing `-swift-version 3` to those.

jrose-apple commented:
Ah, good catch. It probably should.

slavapestov commented:
@dabrahams The standard library needs some minor updates to build in Swift 4 mode, due to the tuple argument changes.

@tkremenek Can you file a bug with the compiler crashers that regressed in Swift 4 mode? They should be rice instead of just passing -swift-version 3.

jrose-apple commented:
@slavapestov We're supposed to be running all the tests under Swift 3 unless the test opts out of that. You can pass "--param swift_version=4" to lit.py to see them running under Swift 4.

tkremenek committed:
Merge branch 'master' into ver-4-bump

tkremenek committed:
Force the benchmark suite to build with -swift-version 3.

Some of the benchmarks use Swift 3 APIs.  Let’s keep them building that
way to not perturb benchmark numbers.

We should consider adding benchmarks that specifically enable -swift-version 4.

tkremenek committed:
These tests explicitly need -swift-version 3.

These tests included sources that are
based on Swift 3.  They can be considered to be updated later to Swift 4, but it did not seem critical for what they were testing.

tkremenek committed:
Update test to reflect new default is -swift-version 4.

tkremenek committed:
Underlying Swift 3 version is now ‘3.2’

tkremenek committed:
baseAddress is optional.

tkremenek commented:
@swift-ci clean test

tkremenek commented:
Please test with following pull request:
https://github.com/apple/swift-xcode-playground-support/pull/14

@swift-ci test macOS

tkremenek committed:
Force specific tests from validation suit to -swift-version 3.

tkremenek commented:
Please test with following pull request:
apple/swift-xcode-playground-support#14

@swift-ci test macOS

jrose-apple commented on test/IDE/complete_crashes.swift:
```diff
@@ -1,7 +1,7 @@
-// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=BAD_MEMBERS_2 | %FileCheck %s -check-prefix=BAD_MEMBERS_2
-// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=CLOSURE_CALLED_IN_PLACE_1 | %FileCheck %s -check-prefix=WITH_GLOBAL
-// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=RDAR_28991372 | %FileCheck %s -check-prefix=RDAR_28991372
+// RUN: %target-swift-ide-test -swift-version 3 -code-completion -source-filename %s -code-completion-token=BAD_MEMBERS_1 | %FileCheck %s -check-prefix=BAD_MEMBERS_1
```
Instead of doing this, can you put `-swift-version 3` into the substitution for %target-swift-ide-test like it is for %target-build-swift?

tkremenek replied:
I tried that and it felt a bit wrong.  Some of these tests will simply not work in both language modes.   I only updated the tests that would not build with the new language default.   That said, I can see the reason why we would want to do it the way you propose so we have the option to run all tests in one mode or the other.

jrose-apple replied:
Whatever we do should be consistent, I think. As long as the swiftc tests are building with -swift-version 3, the swift-ide-test ones should too. Later on we can split up tests that actually should behave differently in Swift 4 mode.

tkremenek replied:
Makes sense to me.

tkremenek committed:
Enhance swift-ide-test to accept -swift-version multiple times.

This allows swift-ide-test to be passed
-swift-version X by lit but have it
overriden by a specifc test.

tkremenek committed:
Add “-swift-version X” to %target-swift-ide-test substitution.

tkremenek committed:
Remove hard passing -swift-version X for some tests.

lit now passes ‘-swift-version X’ to
these tests as part of %target-swift-ide-test.

Note that these tests still require
-swift-version 3.

tkremenek committed:
Resolve merge conflict.

tkremenek commented:
@swift-ci clean test

jrose-apple commented on validation-test/IDE/crashers_2/0010-reference-to-self-in-extension-init.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: not --crash %target-swift-ide-test -code-completion -code-completion-token=A -source-filename=%s
+// RUN: not --crash %target-swift-ide-test -swift-version 3 -code-completion -code-completion-token=A -source-filename=%s
```
This should be implicit now.

jrose-apple commented on validation-test/IDE/crashers_2/0011-complete-protocol-with-initializer.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: not --crash %target-swift-ide-test -code-completion -code-completion-token=A -source-filename=%s
+// RUN: not --crash %target-swift-ide-test -swift-version 3 -code-completion -code-completion-token=A -source-filename=%s
```
This should be implicit now. (Thank you!)

jrose-apple commented on validation-test/Serialization/rdar29694978.swift:
```diff
@@ -1,6 +1,6 @@
-// RUN: %target-build-swift %s -import-objc-header %S/Inputs/rdar29694978.h -emit-module -o %t/Library.swiftmodule
-// RUN: %target-swift-ide-test -print-module -module-to-print=Library -source-filename=x -I %S/Inputs/ -I %t | %FileCheck %s
+// RUN: %target-build-swift %s -swift-version 3 -import-objc-header %S/Inputs/rdar29694978.h -emit-module -o %t/Library.swiftmodule
+// RUN: %target-swift-ide-test -swift-version 3 -print-module -module-to-print=Library -source-filename=x -I %S/Inputs/ -I %t | %FileCheck %s
```
These should be implicit now.

tkremenek replied:
@jrose-apple good catch; forgot those.  I'm re-running my tests locally with those reverted.

jrose-apple reviewed (approved):
Looks pretty good (assuming all our other projects build successfully). Only minor test changes left.

tkremenek commented:
@swift-ci test macOS

tkremenek commented:
Linux still has a bunch of failures that do not show up on macOS:

~~~
17:01:16 Failing Tests (46):
17:01:16     Swift(linux-x86_64) :: Serialization/Recovery/typedefs.swift
17:01:16     Swift(linux-x86_64) :: Constraints/diagnostics.swift
17:01:16     Swift(linux-x86_64) :: ClangImporter/ctypes_parse.swift
17:01:16     Swift(linux-x86_64) :: compiler_crashers_fixed/26813-generic-enum-tuple-optional-payload.swift
17:01:16     Swift(linux-x86_64) :: compiler_crashers_2_fixed/0041-walkToDeclPost-crash.swift
17:01:16     Swift(linux-x86_64) :: compiler_crashers_2_fixed/0034-rdar21982453.swift
17:01:16     Swift(linux-x86_64) :: Sema/type_checker_crashers_fixed/rdar27261929.swift
17:01:16     Swift(linux-x86_64) :: stdlib/Renames.swift
17:01:16     Swift(linux-x86_64) :: expr/postfix/dot/init_ref_delegation.swift
17:01:16     Swift(linux-x86_64) :: expr/closure/single_expr.swift
17:01:16     Swift(linux-x86_64) :: expr/closure/closures.swift
17:01:16     Swift(linux-x86_64) :: expr/closure/anonymous.swift
17:01:16     Swift(linux-x86_64) :: decl/subscript/subscripting.swift
17:01:16     Swift(linux-x86_64) :: decl/protocol/req/optional.swift
17:01:16     Swift(linux-x86_64) :: decl/func/default-values.swift
17:01:16     Swift(linux-x86_64) :: decl/class/override.swift
17:01:16     Swift(linux-x86_64) :: attr/attr_final.swift
17:01:16     Swift(linux-x86_64) :: attr/attr_dynamic.swift
17:01:16     Swift(linux-x86_64) :: Serialization/class-roundtrip-module.swift
17:01:16     Swift(linux-x86_64) :: Serialization/struct.swift
17:01:16     Swift(linux-x86_64) :: Serialization/class.swift
17:01:16     Swift(linux-x86_64) :: Serialization/objc.swift
17:01:16     Swift(linux-x86_64) :: Sema/complex_expressions.swift
17:01:16     Swift(linux-x86_64) :: SILGen/property_abstraction.swift
17:01:16     Swift(linux-x86_64) :: SILGen/properties.swift
17:01:16     Swift(linux-x86_64) :: SILGen/tuples.swift
17:01:16     Swift(linux-x86_64) :: SILGen/partial_apply_generic.swift
17:01:16     Swift(linux-x86_64) :: SILGen/nested_types_referencing_nested_functions.swift
17:01:16     Swift(linux-x86_64) :: SILGen/metatype_abstraction.swift
17:01:16     Swift(linux-x86_64) :: SILGen/function_conversion.swift
17:01:16     Swift(linux-x86_64) :: SILGen/generic_closures.swift
17:01:16     Swift(linux-x86_64) :: SILGen/guaranteed_self.swift
17:01:16     Swift(linux-x86_64) :: SILGen/existential_metatypes.swift
17:01:16     Swift(linux-x86_64) :: SILGen/closures.swift
17:01:16     Swift(linux-x86_64) :: SILGen/dynamic_self.swift
17:01:16     Swift(linux-x86_64) :: SILGen/apply_abstraction_nested.swift
17:01:16     Swift(linux-x86_64) :: Misc/expression_too_complex.swift
17:01:16     Swift(linux-x86_64) :: IRGen/existential_nested_type.swift
17:01:16     Swift(linux-x86_64) :: Generics/deduction.swift
17:01:16     Swift(linux-x86_64) :: DebugInfo/nostorage.swift
17:01:16     Swift(linux-x86_64) :: DebugInfo/WeakCapture.swift
17:01:16     Swift(linux-x86_64) :: DebugInfo/closure-arg-linetable.swift
17:01:16     Swift(linux-x86_64) :: Constraints/tuple.swift
17:01:16     Swift(linux-x86_64) :: Compatibility/attr_escaping.swift
17:01:16     Swift(linux-x86_64) :: Constraints/diag_ambiguities.swift
17:01:16     Swift(linux-x86_64) :: Constraints/closures.swift
~~~

tkremenek committed:
Remove hard-coding of -swift-version 3 in some validation tests.

tkremenek committed:
Pass -swift-version X for %target-swift-frontend

tkremenek commented:
@swift-ci test

tkremenek committed:
Merge branch 'master' into ver-4-bump

tkremenek commented:
@swift-ci clean test

tkremenek committed:
Revert "Pass -swift-version X for %target-swift-frontend"

This reverts commit 157133db50f3fd73377689e5852a6654b21a6c1e.

tkremenek committed:
Include config.swift_test_options in config.target_swift_frontend for non-macOS platforms.

tkremenek commented:
@swift-ci test

tkremenek commented:
On Linux it looks like we are down to a test failure in swift-corelibs-Foundation:

~~~
TestFoundation/HTTPServer.swift:101:20: error: 'subscript' is unavailable
00:31:25             return str[startIndex..<endIndex]
00:31:25                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
00:31:25 Swift.String:3:12: note: 'subscript' was obsoleted in Swift 4
00:31:25     public subscript(bounds: Range<String.Index>) -> String { get }
00:31:25            ^
00:31:25 
TestFoundation/HTTPServer.swift:101:20: error: 'subscript' is unavailable
00:31:25             return str[startIndex..<endIndex]
00:31:25                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
00:31:25 Swift.String:3:12: note: 'subscript' was obsoleted in Swift 4
00:31:25     public subscript(bounds: Range<String.Index>) -> String { get }
00:31:25            ^
~~~

tkremenek commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/972

@swift-ci test linux

tkremenek commented:
Also seeing on iphoneSimulator:

~~~
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/DebugInfo/ErrorVar.swift:19:12: error: expected string not found in input
01:05:25  // CHECK: ![[ERRTY]] = !DICompositeType({{.*}}identifier: "_T0s5Error_pD"
01:05:25            ^
01:05:25 <stdin>:768:1: note: scanning from here
01:05:25 !90 = !DIExpression(DW_OP_deref)
01:05:25 ^
01:05:25 <stdin>:768:1: note: with variable "ERRTY" equal to "85"
01:05:25 !90 = !DIExpression(DW_OP_deref)
01:05:25 ^
01:05:25 <stdin>:773:3: note: possible intended match here
01:05:25 !95 = !DISubroutineType(types: !96)
01:05:25   ^
01:05:25 
~~~

but this looks like a new, unrelated failure.

tkremenek commented:
With:

apple/swift-corelibs-foundation#972

the swift-corelibs-foundation tests pass.  Now I am seeing the following test failure for swift-corelibs-xctest:

~~~
error: expression type 'XCTestCaseEntry' (aka '(testCaseClass: XCTestCase.Type, allTests: Array<(String, (XCTestCase) throws -> ())>)') is ambiguous without more context
06:52:46 XCTMain([testCase(MisuseTestCase.allTests)])
06:52:46          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
06:52:46 
~~~

tkremenek commented:
Please test with following pull request:
apple/swift-corelibs-foundation#972
https://github.com/apple/swift-corelibs-xctest/pull/189

@swift-ci test linux

tkremenek commented:
@swift-ci clean test

tkremenek commented:
I think @adrian-prantl address the `ErrorVar.swift` failure in:

https://github.com/apple/swift/pull/9105

tkremenek commented:
@swift-ci test macOS

tkremenek commented:
I am seeing new test failures that appear unrelated.  Looks like new breakage on `master`.  Merging.

shajrawi commented:
All the linux bots are failing with `static_assert(SWIFT_VERSION_MAJOR == 4` in `swift/lib/Basic/Version.cpp:329`
looking at the build logs I see `  -DSWIFT_VERSION=3.1 -DSWIFT_VERSION_MAJOR=3 -DSWIFT_VERSION_MINOR=1 `

swiftix commented:
@shajrawi You need a clean build to get rid of it

tkremenek commented:
@shajrawi yeah sorry about that; apparently changing the version in the build-script doesn't reflect as a build dependency.  I've triggered the bots to clean their workspaces.

adrian-prantl commented:
AFAIK, this is an inherent limitation of CMake. Once configuration parameters are cached, they can only be reset by removing the CMake cache.

jrose-apple commented:
Maybe we *shouldn't* be marking that variable as "CACHE"?
